core/src/main/java/org/infinispan/notifications/cachelistener/CacheNotifierImpl.java

```diff
@@ -1111,36 +1108,36 @@ public boolean isNotificationAllowed(FlagAffectedCommand cmd, List<CacheEntryLis
    }
 
    private CompletionStage<Void> handlePublisher(CompletionStage<Void> currentStage,
-         Queue<IntermediateOperation> intermediateOperations, QueueingSegmentListener<K, V, ? extends Event<K, V>> handler,
+         Collection<IntermediateOperation<?, ?, ?, ?>> intermediateOperations, QueueingSegmentListener<K, V, ? extends Event<K, V>> handler,
          UUID generatedId, Listener l, Function<Object, Object> kc, Function<Object, Object> kv) {
       SegmentCompletionPublisher<CacheEntry<K, V>> publisher = publisherManager.running().entryPublisher(
             null, null, null, true,
             // TODO: do we really need EXACTLY_ONCE? AT_LEAST_ONCE should be fine I think
             DeliveryGuarantee.EXACTLY_ONCE, config.clustering().stateTransfer().chunkSize(),
             intermediateOperations.isEmpty() ? PublisherTransformers.identity() : new CacheIntermediatePublisher(intermediateOperations));
 
-      io.reactivex.rxjava3.functions.Function<Object, ? extends Publisher<Void>> itemDelayFunction = ice -> {
-         CompletionStage<Void> delay = handler.delayProcessing();
-         if (CompletionStages.isCompletedSuccessfully(delay)) {
-            return Flowable.empty();
-         }
-         return Completable.fromCompletionStage(delay).toFlowable();
-      };
+      io.reactivex.rxjava3.functions.Function<Object, ? extends Publisher<Void>> itemDelayFunction = ice ->
+         RxJavaInterop.voidCompletionStageToFlowable(handler.delayProcessing());
+
       Publisher<CacheEntry<K, V>> p = s -> publisher.subscribe(s, handler);
-      currentStage = Flowable.fromPublisher(p)
-            .delaySubscription(Completable.fromCompletionStage(currentStage).toFlowable())
-            .delay(itemDelayFunction)
-            .filter(ice -> handler.markKeyAsProcessing(ice.getKey()) != QueueingSegmentListener.REMOVED)
-            .delay(ice -> Completable.fromCompletionStage(raiseEventForInitialTransfer(generatedId, ice, l.clustered(),
-                  kc, kv)).toFlowable())
-            // Only request up to 20 at a time
-            .rebatchRequests(20)
-            .count()
-            .toFlowable()
-            // Make sure there are no more delays for processing after we have retrieved all values
-            .delay(itemDelayFunction)
-            .ignoreElements()
-            .toCompletionStage(null);
+
+      currentStage = currentStage.thenCompose(ignore ->
```

Comment:
Wouldn't something like this also work, while being easier to read?

```java
      currentStage = currentStage.thenCompose(ignore ->
            Flowable.fromPublisher(p)
                  .startWith(Completable.defer(() -> Completable.fromCompletionStage(handler.delayProcessing())))
                  .filter(ice -> handler.markKeyAsProcessing(ice.getKey()) != QueueingSegmentListener.REMOVED)
                  .concatMapSingle(ice -> Single.fromCompletionStage(
                        raiseEventForInitialTransfer(generatedId, ice, l.clustered(), kc, kv)))
                  // Only request up to 20 at a time
                  .rebatchRequests(20)
                  // Ignore the actual values
                  .ignoreElements()
                  // Make sure there are no more delays for processing after all the events have been raised
                  .andThen(Completable.defer(() -> Completable.fromCompletionStage(handler.delayProcessing())))
                  .toCompletionStage(null)
      );
```

I'm not 100% sure whether `startWith` is enough, or whether `delay` is still needed at that step. In any case, please add a short comment to explain what it's doing.

Reply:
The `concatMapSingle` would have to be tweaked for sure. I can look into the others though.

Reply:
And `startWith` should work.

Reply:
I have converted all but the `concatMapSingle` since it would need to be
`.concatMapSingle(ice -> Single.fromCompletionStage(raiseEventForInitialTransfer(generatedId, ice, l.clustered(), kc, kv).thenApply(ignore2 -> ice)))`

TBH I wish there was just a `delayStage(Function<Object, CompletionStage<?>)` method or something similar.

Reply:
Right, I didn't realize that `raiseEventForInitialTransfer`'s `CompletionStage` yields `null` so can't be converted to a `Single`.

I do have one small extra request: please move `raiseEventForInitialTransfer` to the next line, the current formatting makes it look like `l.clustered()` is a parameter of `delay()`.

Reply:
Sure.